streaming/replication.c

```diff
@@ -0,0 +1,1931 @@
+                    debug(D_REPLICATION, "%s.%s db empty in valid dimension range @ %ld", st->id, rd->id, metric_t);
+                else {
+                buffer_sprintf(rep_state->build, "FILL \"%s\" \"%s\" %ld " STORAGE_NUMBER_FORMAT "\n", st->id, rd->id, metric_t, n);
+                debug(D_REPLICATION, "%s.%s FILL %ld " STORAGE_NUMBER_FORMAT "\n", st->id, rd->id, metric_t, n);
+                    num_points++;
+                    
+                }
+            }
+            buffer_sprintf(rep_state->build, "FILLEND %zu %u\n", num_points, block_id);
+            rd->state->query_ops.finalize(&handle);
+        }
+        else
+            debug(D_REPLICATION, "%s.%s has no data in the replication window (@%ld-%ld) last_collected=%ld.%ld",
+                                 st->id, rd->id, (long)window_start, window_end, (long)rd->last_collected_time.tv_sec,
+                                 rd->last_collected_time.tv_usec);
+
+    }
+}
+
+void sender_gap_filling(REPLICATION_STATE *rep_state, GAP a_gap)
+{
+    if(!rep_state){
+        error("%s: Cannot Send GAP replication state is NULL!", REPLICATION_MSG);
+        return;
+    }
+    RRDSET *st;
+    RRDHOST *host = rep_state->host;
+    rrdhost_rdlock(host);
+    rrdset_foreach_read(st, host)
+    {
+        sender_chart_gap_filling(st, a_gap); //Send gap data per chart
+    }
+    rrdhost_unlock(host);
+}
+
+// For each chart send the dimension past data in RDATA format
+void sender_chart_gap_filling(RRDSET *st, GAP a_gap) {
+    REPLICATION_STATE *rep_state = st->rrdhost->replication->tx_replication;
+    rrdset_rdlock(st);    
+    if(unlikely(!should_send_chart_matching(st))){
+        rrdset_unlock(st);    
+        return;
+    }
+
+    replication_start(rep_state);         // Locks the sender buffer
+    if(need_to_send_chart_definition(st))
+        replication_send_chart_definition_nolock(st); // Does Replication need to resend the defintions?
+    sender_fill_gap_nolock(rep_state, st, a_gap);
+    replication_commit(rep_state);        // Releases the sender buffer
+    replication_attempt_to_send(rep_state);
+    rrdset_unlock(st);
+}
+
+void replication_gap_to_str(GAP *a_gap, char **gap_str, size_t *len)
+{
+    char gap_uuid_str[UUID_STR_LEN];
+    uuid_unparse(a_gap->gap_uuid, gap_uuid_str);
+    *len = (size_t) (sizeof("GAP ") + UUID_STR_LEN + 3*(10) + 2 + 1);
+    *gap_str = (char *)callocz(*len, sizeof(char));
+    snprintf(*gap_str, *len, "GAP %s %ld %ld %ld\n",
+        gap_uuid_str,
+        a_gap->t_window.t_start,
+        a_gap->t_window.t_first,
+        a_gap->t_window.t_end);
+}
+
+void replication_rdata_to_str(GAP *a_gap, char **rdata_str, size_t *len, int block_id)
+{
+    char gap_uuid_str[UUID_STR_LEN];
+    uuid_unparse(a_gap->gap_uuid, gap_uuid_str);
+    *len = (size_t) (sizeof("RDATA ") + UUID_STR_LEN + 2*(10) + 3 + 1);
+    *rdata_str = (char *)callocz(*len, sizeof(char));
+    snprintf(*rdata_str, *len, "RDATA %s %ld %ld %d\n",
```

Comment:
Considering that the `+1` in *len definition is the null character, here we should have `*len - 1`.
Another thing, please, change `snprintf` to `sprintfz`.